### PR TITLE
Preserve uid gid if running with sudo

### LINKED ISSUE #46854: If `clickhouse` was run as self-extracting executable under sudo, keep the file uid/gid after extracting.
**Use case**

```
sudo ./clickhouse install
```

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Self-extraction with 'sudo' will attempt to set uid and gid of extracted files to running user.

closes #46854

### CODE REVIEW

File: utils/self-extracting-executable/decompressor.cpp [hunk 1]
```diff
@@ -168,6 +168,24 @@ int decompress(char * input, char * output, off_t start, off_t end, size_t max_n
     return 0;
 }
 
+bool getSudoIDs(uid_t &sudo_uid, uid_t &sudo_gid)
```
Comment (alexey-milovidov) [inline utils/self-extracting-executable/decompressor.cpp:171]: This is ok, but... why don't we set it the same as in the original file?
The invariant will be: we don't change the owner/group of the file while extracting.

Comment (yakov-olkhovskiy) [reply utils/self-extracting-executable/decompressor.cpp:171]: yes, makes sense

File: utils/self-extracting-executable/decompressor.cpp [hunk 2]
```diff
@@ -168,6 +168,10 @@ int decompress(char * input, char * output, off_t start, off_t end, size_t max_n
     return 0;
 }
 
+bool isSudo()
+{
+    return getuid() == 0 && geteuid() == 0 && getenv("SUDO_USER") && getenv("SUDO_UID") && getenv("SUDO_GID"); // NOLINT(concurrency-mt-unsafe)
```
Comment (alexey-milovidov) [inline utils/self-extracting-executable/decompressor.cpp:173]: I think geteuid should be enough. We want to chown the file whenever we can.

File: utils/self-extracting-executable/decompressor.cpp [hunk 3]
```diff
@@ -168,6 +168,10 @@ int decompress(char * input, char * output, off_t start, off_t end, size_t max_n
     return 0;
 }
 
+bool isSudo()
+{
+    return geteuid() == 0 && getenv("SUDO_USER") && getenv("SUDO_UID") && getenv("SUDO_GID"); // NOLINT(concurrency-mt-unsafe)
```
Comment (alexey-milovidov) [inline utils/self-extracting-executable/decompressor.cpp:173]: What is the reason for checking SUDO_USER? Doesn't `geteuid` cover all the cases?

Comment (yakov-olkhovskiy) [reply utils/self-extracting-executable/decompressor.cpp:173]: SUDO_ variables make sure that it's sudo and not any other process with setuid - `geteuid` by itself covers too many cases

Comment (alexey-milovidov) [reply utils/self-extracting-executable/decompressor.cpp:173]: Why should we care if it has been run by root without sudo?

Comment (yakov-olkhovskiy) [reply utils/self-extracting-executable/decompressor.cpp:173]: well, the first of all if it was run just by root we want resulting file be owned by root - so no chown is needed here, and the second - it's how the task was set - "If clickhouse was run as self-extracting executable under sudo"... but I mostly agree that it doesn't matter actually, and `geteuid` should be enough

File: utils/self-extracting-executable/decompressor.cpp [hunk 4]
```diff
@@ -168,6 +168,10 @@ int decompress(char * input, char * output, off_t start, off_t end, size_t max_n
     return 0;
 }
 
+bool isSudo()
```
Comment (alexey-milovidov) [inline utils/self-extracting-executable/decompressor.cpp:171]: Now the function name became misleading...

But LGTM, let's merge :)

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**